### ISSUE TITLE
BUG - Variable Compile and Load to EEPROM error with Variables - Live BlocklyProp Beta

### ISSUE THREAD

**Issue by wstodghill** [parallaxinc/BlocklyProp#1058]
Trying to compile or load to EEPROM for the following code using a variable in the Live BlocklyProp Beta site...
![variable error](https://user-images.githubusercontent.com/8891440/27465827-a0fcc340-579b-11e7-9be6-8423512e199e.JPG)

yields the following error message (similar for compile).

![error message](https://user-images.githubusercontent.com/8891440/27465839-b55b4208-579b-11e7-82e1-46af9754a773.JPG)

Sorry, I am finding all these! Been using the DEMO site so long - I have not been having any issues.

**Comment by wstodghill** [comment]
Oddly enough, in this very similar example - everything works fine!
See - http://blockly.parallax.com/blockly/editor/blocklyc.jsp?project=8038

![uh oh](https://user-images.githubusercontent.com/8891440/27466243-c94d701c-579e-11e7-9253-684cbe987725.JPG)

**Comment by wstodghill** [comment]
@kgracey, @MatzElectronics, @zfi, - please take a look.

**Comment by zfi** [comment]
The blocks in the first diagram generate the following code and also does trigger an error in the Cloud Compiler service:

```
// ------ Libraries and Definitions ------
#include "simpletools.h"
#include "s3.h"

// ------ Global Variables and Objects ------
int time;



// ------ Main Program ------
int main()
{
    s3_setup();
    pause(100);
    time = 500;
    s3_setLED(S3_LEFT, S3_COLOR_00FF00);
    s3_setLED(S3_CENTER, S3_COLOR_00FF00);
    s3_setLED(S3_RIGHT, S3_COLOR_00FF00);
    pause(time);
    s3_setLED(S3_LEFT, S3_COLOR_000000);
    s3_setLED(S3_CENTER, S3_COLOR_FF0000);
    s3_setLED(S3_RIGHT, S3_COLOR_000000);

}
```

However, if the variable 'time' is renamed to 'time1', the resulting code (shown below) does compile cleanly. 

```
// ------ Libraries and Definitions ------
#include "simpletools.h"
#include "s3.h"

// ------ Global Variables and Objects ------
int time1;



// ------ Main Program ------
int main()
{
    s3_setup();
    pause(100);
    time1 = 500;
    s3_setLED(S3_LEFT, S3_COLOR_00FF00);
    s3_setLED(S3_CENTER, S3_COLOR_00FF00);
    s3_setLED(S3_RIGHT, S3_COLOR_00FF00);
    pause(time1);
    s3_setLED(S3_LEFT, S3_COLOR_000000);
    s3_setLED(S3_CENTER, S3_COLOR_FF0000);
    s3_setLED(S3_RIGHT, S3_COLOR_000000);

}
```

One of the libraries is referencing the time structure defined in time.h within the library and this is colliding with the user-land variable 'time'.

**Comment by zfi** [comment]
Worked with ALindsay to understand the issue. There are a number of related errors possible whenever a user introduces a variable that is the same as the name of any included library function name or library-defined global variables.

Need to discuss with @MatzElectronics to workout a resolution to this.

In the interim, the recommendation is to change the user-land variable name to something that does not generate the conflict.

**Comment by wstodghill** [comment]
@zfi said, "In the interim, the recommendation is to change the user-land variable name to something that does not generate the conflict."

Patient - "Doctor, it hurts when I do this."
Doctor - "Don't do that."

;-)

**Comment by wstodghill** [comment]
If I haven't said it enough lately - I am glad smarter people than me are on the case! Thanks for the help!

**Comment by MatzElectronics** [comment]
The solution I would prefer to implement is to use is for the variable/function block to display a warning when the user tries to type in a variable name that will collide with an existing variable/function.  This is easy to implement, if I have a list of those variables/functions/reserved words.

I just worked out an initial list that should cover almost all of the c reserved words as well as many of the function names from propeller.h, simpletext.h, and simpletools.h.

We will need to scan through the remainder of the .h files in Simple Libraries to catch them all - but this should get most of the collisions we are seeing right now.

**Comment by wstodghill** [comment]
All,
Okay. Duh - just re-read this... []]((http://learn.parallax.com/support/reference/propeller-blocklyprop-block-reference/about-naming-things) This is the issue, yes?

**Comment by zfi** [comment]
The article does note the C language reserved words, but it does not talk about global variables and function names used in the libraries. 

For example, there is a pause() function in one of the libraries. If a project is using a variable named pause, it works unless the project is also using some of the delay blocks. That's what happened with the time variable noted at the top of this issue.

**Comment by wstodghill** [comment]
@zfi - Thanks for the clarification! You all are the greatest.

**Comment by MatzElectronics** [comment]
This is caused but the code intended to fix #1016.  Closing this issue and addressing it there.
